README
Libbraille - Library to easily access Braille displays and terminals

Web Page: http://libbraille.org

Libbraille comes with ABSOLUTELY NO WARRANTY.

This is free software, placed under the terms of the GNU Lesser
General Public License, as published by the Free Software Foundation.
Please see the file COPYING for details.

Currently supported hardware
============================

The following Braille displays are supported :

* Alva ABT320, ABT340, ABT340 Desktop, ABT380, ABT380 Twin Space,
Delphi 40, Delphi 80, Satellite 544, 570 Pro, 584 Pro and 544
Traveller
* Baum Pocket and super Vario 24, 32, 40, 64 and 80
* Blazie Engineering BrailleLite 18, 40, Millenium 20 and Millenium 40
* EuroBraille NoteBRAILLE, Clio-NoteBRAILLE, SCRIBA, AzerBRAILLE,
CLIO-euroBraille and Iris
* Handytech 40, 80 and BrailleWave
* Hermes
* MDV MB 408S
* ONCE Eco 20, 40 and 80
* Papenmeier Braillex EL 40S
* Pulse Data BrailleNote 18 and 32
* TechniBraille Manager 40i and 40s
* Tieman Combibraille 45 and 85
* Tieman Voyager 44, 44+, 70 and 70+


ChangeLog
2007-08-17  Sébastien Sablé  <sable@users.sourceforge.net>

	* backend/voyager/brl.c: Corrected memory leak when closing
	voyager driver

2006-03-02  Sébastien Sablé  <sable@users.sourceforge.net>

	* Release 0.19.0
	* scripts/libbraille.inf: Updated for libusb-win32 20051120
	* lib/braille.c: Corrected bug in braille_write where extended
	ASCII char are discarded (thanks to Ismo Salonen)
	* Updated copyright headers
	* Various cleaning before releasing 0.19.0

2005-10-01  Sébastien Sablé  <sable@users.sourceforge.net>

	* Rewrote Baum serial driver using documentation provided by
	Baum. Added support for USB devices.

2005-09-30  Sébastien Sablé  <sable@users.sourceforge.net>

	* All backends: passing a type argument to brli_drvinit
	* Improved autodetection system
	* Various minor cleanings
	* lib/serial_win32.c, lib/serial_tios.c: Removed deprecated old
	serial API
	* backend/eurobraille/brl.c: Added support for additionnal Scriba
	model
	* scripts/libbraille.inf: Added USB IDs for Papenmeier display

2005-09-01  Sébastien Sablé  <sable@users.sourceforge.net>

	* backend/papenmeierusb: Added driver for Papenmeier USB displays
	thanks to a loan by Alphabraille and minimal documentation from
	Papenmeier
	* backend/alvausb/brl.c: Corrected Alvausb driver to make it work
	on Mac

2005-08-31  Sébastien Sablé  <sable@users.sourceforge.net>

	* backend/fakegtk/brl.c: Do not call gtk_main if the library is
	called from a Gtk+ application (necessary for Mozilla for example)
	* backend/fakegtk/brl.c: Updated to use asynchronous queue instead
	of locks to wait between threads
	* backend/voyager/brl.c: Corrected Voyager driver to make it work
	on Mac
	* backend/alva: Rewrote the Alva serial driver with new serial
	API, autodetection and timeout handling thanks to documentation
	provided by Alva
	* backend/eurobraille: Added detection of the CLIO-eurobraille
	display in eurobraille driver
	* lib/config_*.c: Updated list of tables
	* test/test_libbraille.c: Add output of debug messages to the
	Braille display when pressing keys
	* lib/config.c: Autodetect usb displays first then serial displays
	* lib/braille.c: Lower usb debug log
	* Makefile.am: Cleaning
	* lib/serial_tios.c: Mac portability - use usleep instead of
	select. Do not set serial output speed at 0 on init
	* configure.ac, lib/Makefile.am: Improved libusb detection
	* autogen_mac.sh: Adding Mac specific autogen script 

2004-11-26  Sébastien Sablé  <sable@users.sourceforge.net>

	* Release 0.18.0
	* Added win32 Python module to NSIS windows installer
	* Compiling win32 Python module with visual studio

2004-11-24  Sébastien Sablé  <sable@users.sourceforge.net>

	* Rewrote the handytech driver with new serial API, autodetection
	and timeout handling

2004-11-23  Sébastien Sablé  <sable@users.sourceforge.net>

	* Unified log message at detection

2004-11-22  Sébastien Sablé  <sable@users.sourceforge.net>

	* Added HTML documentation to RPM devel package
	* Added HTML documentation to debian devel package
	* Added --enable-doc option to configure to generate HTML
	documentation

2004-11-19  Sébastien Sablé  <sable@users.sourceforge.net>

	* Created libbraille-fake debian package for virtual display

2004-11-18  Sébastien Sablé  <sable@users.sourceforge.net>

	* Corrected compilation with Mingw32 compiler

2004-11-15  Sébastien Sablé  <sable@users.sourceforge.net>

	* Added documentation to windows installer
	* Corrected timeout in 'none' driver
	* Check for libusb >= 0.1.8
	* Split config.c in config_posix.c and config_win32.c

2004-11-13  Sébastien Sablé  <sable@users.sourceforge.net>

	* Passing a brli_term* argument to brli_drvinfo
	* Created a RPM package of libbraille

2004-11-09  Sébastien Sablé  <sable@users.sourceforge.net>

	* braille_info(BRL_PATH) returns installation prefix in unix and installation
	location in windows. BRL_PATHCONF does not exist in windows and
	corresponds to the directory where is the configuration file in unix.

2004-11-08  Sébastien Sablé  <sable@users.sourceforge.net>

	* Release 0.17.0

2004-11-05  Sébastien Sablé  <sable@users.sourceforge.net>

	* Split BRL_PATH option in BRL_PATHCONF where to look for
	configuration file in unix, BRL_PATHDRV where to look for drivers
	and BRL_PATHTBL where to look for tables
	* Split serial.c in serial_tios.c for termios and serial_win32.c
	for win32

2004-11-04  Sébastien Sablé  <sable@users.sourceforge.net>

	* Updated Fake GTK virtual display to handle timeouts like other
	drivers
	* Corrected thread problem which prevented correct refreshing in
	the Fake Gtk+ virtual display driver

2004-11-03  Sébastien Sablé  <sable@users.sourceforge.net>

	* Added driver for Technibraille displays
	* Corrected timeout problem in serial communication
	* Added braille_debug function to choose the level of verbosity
	* Corrected compilation warnings

2004-10-05  Sébastien Sablé  <sable@users.sourceforge.net>

	* Release 0.16.0
	* Changed test script to pause before exiting in win32 and to quit
	after a few seconds

2004-10-04  Sébastien Sablé  <sable@users.sourceforge.net>

	* Added driver for Tieman Voyager USB display thanks to doc from
	Tieman
	* Added usb driver to windows installer
	* Changed win32 directories structure for installation system

2004-09-09  Sébastien Sablé  <sable@users.sourceforge.net>

	* Added a windows installer using NSIS http://nsis.sf.net
	* Rewrote configuration system using the registry in win32

2004-08-30  Sébastien Sablé  <sable@users.sourceforge.net>

	* Release 0.15.0

2004-08-26  Sébastien Sablé  <sable@users.sourceforge.net>

	* Correcting small time problem in blocking mode under win32
	* Filtering invalid ascii chars in braille_write and
	braille_statuswrite
	* Removed useless BRAILLECALL calling convention modifier
	* Added a function to get the type of a driver (special, serial or
	usb)
	* Updated debian package including support for hotplug
	* Separated library version from package version for better library versioning


lib/config.c
/* Libbraille - a portable library for Braille displays
 * Copyright (C) 2001-2006 by Sébastien Sablé
 * 
 * This program comes with ABSOLUTELY NO WARRANTY.
 *
 * This is free software, placed under the terms of the GNU Lesser
 * Library General Public License, as published by the Free Software
 * Foundation.  Please see the file COPYING for details.  */

#include "config.h"
#include "brl_config.h"
#include "brl_error.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

extern brl_table table_list[];

static brl_driver driver_list[] =
  {
    {"auto",          0, BRL_TYPE_SPECIAL,
     "Autodetect the type of terminal"},
    {"alva",          1, BRL_TYPE_SERIAL,
     "Alva ABT320, ABT340, ABT340 Desktop, ABT380, ABT380 Twin Space, "
     "Delphi 420, 440, 440 Desktop, 480, Satellite 544, 570 Pro, 584 Pro and 544 Traveller"},
    {"alvausb",       1, BRL_TYPE_USB,
     "Alva Satellite 544, 570 Pro, 584 Pro and 544 Traveller"},
    {"baum",          1, BRL_TYPE_USB | BRL_TYPE_SERIAL,
     "Baum Pocket and Super Vario 24, 32, 40, 64 and 80"},
    {"braillelite",   1, BRL_TYPE_SERIAL,
     "BrailleLite 18, 40, Millenium 20 and Millenium 40"},
    {"braillenote",   1, BRL_TYPE_SERIAL,
     "Pulse Data BrailleNote 18 and 32"},
    {"eurobraille",   1, BRL_TYPE_SERIAL,
     "NoteBRAILLE, Clio-NoteBRAILLE, SCRIBA, AzerBRAILLE, CLIO-euroBraille and Iris"},
    {"fake",          0, BRL_TYPE_SPECIAL,
     "Fake"},
    {"handytech",     1, BRL_TYPE_SERIAL,
     "HandyTech 40, 80, Braille Wave, Baum Vario 40 via emulation"},
    {"none",          0, BRL_TYPE_SPECIAL, "None"},
    {"once",          1, BRL_TYPE_SERIAL,
     "ONCE Eco 20, 40 and 80"},
    {"papenmeierusb", 1, BRL_TYPE_USB, "Papenmeier Braillex EL 40S"},
    {"technibraille", 1, BRL_TYPE_SERIAL, "Manager 40i and 40s"},
    {"text",          0, BRL_TYPE_SPECIAL, "Text"},
    {"voyager",       1, BRL_TYPE_USB, "Tieman Voyager 44, 44+, 70 and 70+"},
  };

/* Returns the number of available drivers */
BRAILLE_API int
braille_drivernum(void)
{
  return sizeof(driver_list)/sizeof(brl_driver);
}

/* Returns the name of the driver for which the number is provided */
BRAILLE_API const char *
braille_drivername(int num)
{
  if(num < 0 || num > sizeof(driver_list)/sizeof(brl_driver))
    return NULL;

  return driver_list[num].name;
}

/* Returns the type of the driver (usb or serial) for which the number
   is provided */
BRAILLE_API brl_drivertype
braille_drivertype(int num)
{
  if(num < 0 || num > sizeof(driver_list)/sizeof(brl_driver))
    return 0;

  return driver_list[num].type;
}

/* Returns a string describing models supported by the driver for
   which the number is provided */
BRAILLE_API const char *
braille_drivermodels(int num)
{
  if(num < 0 || num > sizeof(driver_list)/sizeof(brl_driver))
    return NULL;

  return driver_list[num].models;
}

/* Returns whether a driver can autodetect its terminal */
BRAILLE_API char
braille_driverauto(int num)
{
  if(num < 0 || num > sizeof(driver_list)/sizeof(brl_driver))
    return 0;
  
  return driver_list[num].autodetect;
}

/* Returns whether a driver supports a specific connection type */
BRAILLE_API char
braille_driver_is_type(int num, char type)
{
  if(num < 0 || num > sizeof(driver_list)/sizeof(brl_driver))
    return 0;
  
  return (driver_list[num].type & type ? 1 : 0);
}

/* Configure some values */
BRAILLE_API char
braille_config(brl_config code, char *value)
{
  size_t len;

  switch(code)
    {
    case BRL_DEVICE:
      if(device)
	{
	  free(device);
	}
      if(value == NULL)
	{
	  device = NULL;
	  break;
	}
      len = strlen(value);
      device = (char *)malloc(len + 1);
      if(!device)
	{
	  brli_seterror("Error allocating memory:%s", strerror(errno));
	  return 0;
	}
      strncpy(device, value, len + 1);
      break;
    case BRL_DRIVER:
      if(driver)
	{
	  free(driver);
	}
      if(value == NULL)
	{
	  driver = NULL;
	  break;
	}
      len = strlen(value);
      driver = (char *)malloc(len + 1);
      if(!driver)
	{
	  brli_seterror("Error allocating memory:%s", strerror(errno));
	  return 0;
	}
      strncpy(driver, value, len + 1);
      break;
    case BRL_TABLE:
      if(table)
	{
	  free(table);
	}
      if(value == NULL)
	{
	  table = NULL;
	  break;
	}
      len = strlen(value);
      table = (char *)malloc(len + 1);
      if(!table)
	{
	  brli_seterror("Error allocating memory:%s", strerror(errno));
	  return 0;
	}
      strncpy(table, value, len + 1);
      break;
    case BRL_PATH:
#ifdef WIN32
      if(path)
	{
	  free(path);
	}
      if(value == NULL)
	{
	  path = NULL;
	  break;
	}
      len = strlen(value);
      path = (char *)malloc(len + 1);
      if(!path)
	{
	  brli_seterror("Error allocating memory:%s", strerror(errno));
	  return 0;
	}
      strncpy(path, value, len + 1);
      break;
#else
      break;
#endif
#ifndef WIN32
    case BRL_PATHCONF:
      if(pathcfg)
	{
	  free(pathcfg);
	}
      if(value == NULL)
	{
	  pathcfg = NULL;
	  break;
	}
      len = strlen(value);
      pathcfg = (char *)malloc(len + 1);
      if(!pathcfg)
	{
	  brli_seterror("Error allocating memory:%s", strerror(errno));
	  return 0;
	}
      strncpy(pathcfg, value, len + 1);
      break;
#endif
    case BRL_PATHDRV:
      if(pathdrv)
	{
	  free(pathdrv);
	}
      if(value == NULL)
	{
	  pathdrv = NULL;
	  break;
	}
      len = strlen(value);
      pathdrv = (char *)malloc(len + 1);
      if(!pathdrv)
	{
	  brli_seterror("Error allocating memory:%s", strerror(errno));
	  return 0;
	}
      strncpy(pathdrv, value, len + 1);
      break;
    case BRL_PATHTBL:
      if(pathtbl)
	{
	  free(pathtbl);
	}
      if(value == NULL)
	{
	  pathtbl = NULL;
	  break;
	}
      len = strlen(value);
      pathtbl = (char *)malloc(len + 1);
      if(!pathtbl)
	{
	  brli_seterror("Error allocating memory:%s", strerror(errno));
	  return 0;
	}
      strncpy(pathtbl, value, len + 1);
      break;
    default:
      brli_seterror("Unrecognised configuration code");
      return 0;
    }
  return 1;
}
